nyxt: fix darwin build

midchildan opened the pull request:
<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

###### Motivation for this change
Continuation of #125113.

* Fix the Darwin build
* Use the GI-GTK backend as recommended by @jmercouris
* Ran nixpkgs-fmt

Keyboard input is still not working, but it turned out to be an issue with WebKitGTK so I'm preparing a separate issue for it. Currently, all packages using WebKitGTK would have the same issue on Darwin.

ping @mjlbach

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [x] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

midchildan committed:
nyxt: fix darwin build

Co-authored-by: Michael Lingelbach <m.j.lbach@gmail.com>

midchildan commented on pkgs/applications/networking/browsers/nyxt/default.nix:
```diff
@@ -12,36 +22,62 @@ stdenv.mkDerivation rec {
-  ] ++ gstBuildInputs;
+  ]);
 
-  GST_PLUGIN_SYSTEM_PATH_1_0 = lib.concatMapStringsSep ":" (p: "${p}/lib/gstreamer-1.0") gstBuildInputs;
```
Setting `GST_PLUGIN_SYSTEM_PATH_1_0` is handled by `wrapGApps`.  It suffices to put these in `buildInputs`.

midchildan commented on pkgs/applications/networking/browsers/nyxt/default.nix:
```diff
@@ -12,36 +22,62 @@ stdenv.mkDerivation rec {
-    glib gdk-pixbuf cairo
-    mime-types pango gtk3
-    glib-networking gsettings-desktop-schemas
-    xclip notify-osd enchant
```
`xclip` in `buildInputs` doesn't have any effect because it's a CLI tool, not a library. I've added `xclip` to the wrapper instead.

midchildan replied:
`notify-osd` is removed since it's no longer needed. Details in  https://github.com/NixOS/nixpkgs/pull/125113#issue-658460773.

midchildan commented on pkgs/applications/networking/browsers/nyxt/default.nix:
```diff
@@ -12,36 +22,62 @@ stdenv.mkDerivation rec {
-      cp -f $src/lib/common-lisp/nyxt/assets/nyxt_''${i}x''${i}.png "$out/share/icons/hicolor/''${i}x''${i}/apps/nyxt.png"
-    done
-
-    mkdir -p $out/bin && makeWrapper $src/bin/nyxt $out/bin/nyxt \
```
This didn't work because `gappsWrapperArgs` is computed at `preFixup`, which happens after `installPhase`.

midchildan commented on pkgs/development/lisp-modules/lisp-packages.nix:
```diff
@@ -129,22 +129,42 @@ let lispPackages = rec {
+
       postInstall = ''
         echo "Building nyxt binary"
-        (
-          source "$out/lib/common-lisp-settings"/*-shell-config.sh
-          cd "$out/lib/common-lisp"/*/
-          makeFlags="''${makeFlags:-}"
-          make LISP=common-lisp.sh NYXT_INTERNAL_QUICKLISP=false PREFIX="$out" $makeFlags all
-          make LISP=common-lisp.sh NYXT_INTERNAL_QUICKLISP=false PREFIX="$out" $makeFlags install
-          cp nyxt "$out/bin/nyxt"
-        )
```
This was redundant because it creates a broken build artifact `$out/bin/nyxt` which promptly gets rebuilt by the subsequent commands.

midchildan commented on pkgs/development/lisp-modules/lisp-packages.nix:
```diff
@@ -129,22 +129,42 @@ let lispPackages = rec {
-           "(asdf/system:component-entry-point (asdf:find-system :nyxt/gtk-application))" \
-           "" "(format *error-output* \"Alien objects:~%~s~%\" sb-alien::*shared-objects*)"
+          nix_lisp_build_system nyxt/gi-gtk-application \
+            "(asdf/system:component-entry-point (asdf:find-system :nyxt/gi-gtk-application))" \
+            "" \
+            "(format *error-output* \"Alien objects:~%~s~%\" sb-alien::*shared-objects*)"
         ' "$out/bin/nyxt-lisp-launcher.sh"
-        cp "$out/lib/common-lisp/nyxt/nyxt" "$out/bin/"
+
+        mv "$out/lib/common-lisp/nyxt/nyxt" "$out/bin/"
```
Replaces the GTK backend with the GI-GTK one.

midchildan commented on pkgs/development/lisp-modules/lisp-packages.nix:
```diff
@@ -129,22 +129,42 @@ let lispPackages = rec {
-        )
+
+        # clear unnecessary environment variables to avoid hitting the limit
+        env -i \
```
It's necessary to clear unnecessary environment variables because otherwise it hits the length limit imposed by the OS. This is similar to what is done in `lisp-modules/define-package.nix`:
https://github.com/NixOS/nixpkgs/blob/8284fc30c84ea47e63209d1a892aca1dfcd6bdf3/pkgs/development/lisp-modules/define-package.nix#L89-L99

r-rmcgibbo commented:
Result of `nixpkgs-review pr 126082` at 0498cf00 run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package failed to build:</summary>
  <ul>
    <li><a href="https://gist.github.com/c25d25a337900cea64877ce653b2416b">nyxt</a></li>
  </ul>
</details>
<details>
  <summary>1 package built successfully:</summary>
  <ul>
    <li>lispPackages.nyxt</li>
  </ul>
</details>
<details>
  <summary>2 suggestions:</summary>
  <ul>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/missing-phase-hooks.md">missing-phase-hooks</a>

`checkPhase` should probably contain `runHook preCheck` and `runHook postCheck`.

Near pkgs/applications/networking/browsers/nyxt/default.nix:82:3:
```
   |
82 |   checkPhase = ''
   |   ^
```
</li>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/missing-phase-hooks.md">missing-phase-hooks</a>

`installPhase` should probably contain `runHook preInstall` and `runHook postInstall`.

Near pkgs/applications/networking/browsers/nyxt/default.nix:51:3:
```
   |
51 |   installPhase =
   |   ^
```
</li>
  </ul>
</details>

Note that build failures [may predate](https://github.com/nix-community/hydra-check) this PR, and could be nondeterministic or hardware dependent.
Please exercise your independent judgement. Does something look off? Please file an [issue](https://github.com/rmcgibbo/r-rmcgibbo/issues) or reach out on IRC.

midchildan commented:
I've pushed a few more fixes that does the following:

* Add `dontStrip = true` to `nyxt` to fix the Linux build. Stripping results in a crash, likely because the resulting binary is already stripped once in `lispPackages.nyxt`.
* Actually run the `checkPhase`, and run make it `installCheckPhase` because it needs to run after `installPhase`. I've also disabled it for Darwin because the test would initialize the GTK runtime, which I suspect would fail on Hydra.
* Add `runHook` as suggested by @r-rmcgibbo

babariviere commented:
Result of `nixpkgs-review pr 126082` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages built:</summary>
  <ul>
    <li>lispPackages.nyxt</li>
    <li>nyxt</li>
  </ul>
</details>

mjlbach commented:
The changes look good to me:

* As noted, keyboard input event don't work unless holding down `cmd`, which may be a webkitgtk issue
* It's a bit crash-ey for me, I can't use it more than 5 minutes without a crash, but that's probably an upstream issue
* It's still quite laggy

midchildan committed:
lispPackages.nyxt-unwrapped: rename from lispPackages.nyxt

midchildan commented:
I've now renamed `lispPackages.nyxt` to `lispPackages.nyxt-unwrapped` for more clarity and updated the release notes accordingly. This should have minimal effect because most users would be have `nyxt` installed instead.

Also rebased to fix the merge conflict in the release notes.

SuperSandro2000 commented on nixos/doc/manual/from_md/release-notes/rl-2111.section.xml:
```diff
@@ -64,6 +64,12 @@
           this version for the entire lifecycle of the 21.11 release.
         </para>
       </listitem>
+      <listitem>
+        <para>
+          <literal>lispPackages.nyxt</literal> has been renamed to
+          <literal>lispPackages.nyxt-unwrapped</literal>
+        </para>
+      </listitem>
```
Just add an alias to aliases.nix instead.

midchildan replied:
It wasn't possible to add it because `lispPackages.nyxt` isn't a top-level package, so I made `lispPackages.nyxt` to throw an error message instead.

ethancedwards8 commented:
Result of `nixpkgs-review pr 126082` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>nyxt</li>
  </ul>
</details>
